File: backend/src/main/java/com/leavemarker/entity/LeaveApplication.java
package com.leavemarker.entity;

import com.leavemarker.enums.LeaveStatus;
import com.leavemarker.enums.LeaveType;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;

@Entity
@Table(name = "leave_applications")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class LeaveApplication extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "employee_id", nullable = false)
    private Employee employee;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private LeaveType leaveType;

    @Column(nullable = false)
    private LocalDate startDate;

    @Column(nullable = false)
    private LocalDate endDate;

    @Column(nullable = false)
    private Double numberOfDays;

    @Column(nullable = false)
    private Boolean isHalfDay = false;

    @Column(length = 1000)
    private String reason;

    @Column(length = 500)
    private String attachmentUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private LeaveStatus status = LeaveStatus.PENDING;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by_manager_id")
    private Employee approvedByManager;

    @Column
    private LocalDate managerApprovalDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by_hr_id")
    private Employee approvedByHr;

    @Column
    private LocalDate hrApprovalDate;

    @Column(length = 500)
    private String rejectionReason;

    @Column
    private LocalDate rejectionDate;

    @Column(nullable = false)
    private Boolean requiresHrApproval = false;
}

File: backend-go/internal/models/leave_application.go
package models

import (
	"time"

	"github.com/leavemarker/backend-go/internal/enums"
)

// LeaveApplication represents a leave application entity
type LeaveApplication struct {
	BaseModel
	EmployeeID          uint              `gorm:"not null;index" json:"employeeId"`
	LeaveType           enums.LeaveType   `gorm:"size:30;not null" json:"leaveType"`
	StartDate           time.Time         `gorm:"type:date;not null" json:"startDate"`
	EndDate             time.Time         `gorm:"type:date;not null" json:"endDate"`
	NumberOfDays        float64           `gorm:"not null" json:"numberOfDays"`
	IsHalfDay           bool              `gorm:"default:false" json:"isHalfDay"`
	Reason              string            `gorm:"size:1000" json:"reason"`
	AttachmentURL       *string           `gorm:"size:500" json:"attachmentUrl"`
	Status              enums.LeaveStatus `gorm:"size:20;default:PENDING" json:"status"`
	ApprovedByManagerID *uint             `gorm:"index" json:"approvedByManagerId"`
	ManagerApprovalDate *time.Time        `gorm:"type:date" json:"managerApprovalDate"`
	ApprovedByHRID      *uint             `gorm:"index" json:"approvedByHrId"`
	HRApprovalDate      *time.Time        `gorm:"type:date" json:"hrApprovalDate"`
	RejectionReason     *string           `gorm:"size:500" json:"rejectionReason"`
	RejectionDate       *time.Time        `gorm:"type:date" json:"rejectionDate"`
	RequiresHRApproval  bool              `gorm:"default:false" json:"requiresHrApproval"`

	// Relationships
	Employee          Employee  `gorm:"foreignKey:EmployeeID" json:"-"`
	ApprovedByManager *Employee `gorm:"foreignKey:ApprovedByManagerID" json:"-"`
	ApprovedByHR      *Employee `gorm:"foreignKey:ApprovedByHRID" json:"-"`
}

func (LeaveApplication) TableName() string {
	return "leave_applications"
}
